Custom\Manual waveguide mode addition, for arbitrary waveguide structures

gadiLhv opened the pull request:
This one was rather tricky. Took me a good long while to figure out what functionality I should access in openEMS. 

___
I'll start from the tricky part, in `ports.py`. This is where openEMS and CSXCAD are intertwined. I needed some sort of validation of the inputs, to add this functionality to `WaveguidePort`. Until the CSXCAD PR isn't finalized, this can't be, either. Unless you want to neglect the python functionality for now.
https://github.com/thliebig/openEMS/compare/master...gadiLhv:openEMS:WaveguideModes?plain=1#diff-fe6a56ae621157b6691266fcc774ca626b01b38cb611a5b21e623b9a9d3493c7R382-R391
___

The only bit of redundancy may be this:
https://github.com/thliebig/openEMS/compare/master...gadiLhv:openEMS:WaveguideModes#diff-53e7f365700486b254789d1b8ff10d0073b4bfd6b44e4b5823f72f2192d1b96fR282-R305

where this doesn't exist anywhere else in openEMS, rather in CSXCAD. If I understand correctly, you would rather some sort of field file handling class for this. 
___

Here is the clever bit, though. I didn't write anything to the engine extension. It's actually done in the `GetWeightedExcitation` function, here:
https://github.com/gadiLhv/openEMS/blob/4195be2c83c118ea9f26ddf8eff90715744d7a68/FDTD/extensions/operator_ext_excitation.cpp#L205

and here:
https://github.com/gadiLhv/openEMS/blob/4195be2c83c118ea9f26ddf8eff90715744d7a68/FDTD/extensions/operator_ext_excitation.cpp#L260C9-L260C84

gadiLhv committed:
Compiles, Runs and ready for PR
Added test scripts

gadiLhv committed:
Started integration of mode parsers

Not finished
Far from compiling.

TODO:
1. Finish probes
2. Verify input validity
3. Try to find more elegant solution for excitation.

gadiLhv committed:
Forgot to approve processmodematch.h

gadiLhv committed:
Fixed coordinate shifting in processmodematch

gadiLhv committed:
Compiles, untested!

1. Finished mode match integration
2. Fixed compilation errors
3. Changed python bindings to work with file name

gadiLhv committed:
Fixed ports.py

gadiLhv committed:
Debugging operator_ext_exciation

gadiLhv committed:
Debugged excitation modes

1. Issues in ports.py
2. Minor changes to operator_ext_exciation

TODO: Still need to validate everything and check probes & modes modes.

gadiLhv committed:
Fixed final bugs in ports.py

1. Fixed direction and naming bugs in ports.py
2. Added example script and example mode files

gadiLhv commented:
Hey @thliebig 

What I couldn't (easily) overcome here was in the excitation extension, I couldn't pass just the file name.
This is due to the fact that the excitation initialization is done throughout the grid, rather than primitive-by-primitive.
This forced the file parsing to be done within the CSPropExcitation itself, but it's done with the file parse, while the only input given is the file name.

Other than that, pretty straight forward. Hope I didn't mess anything else up, too bad.

Cheers

gadiLhv committed:
Changed example file folder location

gadiLhv committed:
Changed Example file folder location

gadiLhv commented:
Hey @thliebig 

First of all, thank you for merging the local obsorber PR. I hope it will be useful for other things, as the SIBC/MUR-SA is a pretty cool addition.

This PR went through rather extensive changes (along with it's CSXCAD counterpart) so let me know how do you find it.

Cheers

thliebig commented on python/openEMS/ports.py:
```diff
@@ -347,54 +347,108 @@ class WaveguidePort(Port):
     """
-    def __init__(self, CSX, port_nr, start, stop, exc_dir, E_WG_func, H_WG_func, kc, excite=0, **kw):
-        super(WaveguidePort, self).__init__(CSX, port_nr=port_nr, start=start, stop=stop, excite=excite, **kw)
+    def __init__(self, CSX, port_nr, start, stop, exc_dir, E_WG_func, H_WG_func, kc, excite=0, excite_type=0, **kw):
```
I'm not a fan of reusing the E/H_WG_func as file name redefinition...
Better set these two to None and have a new argument for the file input instead?
If kc gets a default argument too then all E/H_WG_func and kc could be set to None by default and evaluated as a correct combination more easily and more readable?

thliebig commented on python/openEMS/ports.py:
```diff
@@ -347,54 +347,108 @@ class WaveguidePort(Port):
         self.direction = np.sign(stop[self.exc_ny]-start[self.exc_ny])
         self.ref_index = 1
-
+        
```
Please avoid trailing whites-spaces... check if you can configure your editor?

gammaxy1 replied:
Lines 358 and 437 were both empty before and now have 8 spaces.

thliebig replied:
Okay maybe trailing was a bit misleading, empty lines should not have white-spaces too, just empty please.

thliebig commented on python/openEMS/ports.py:
```diff
@@ -347,54 +347,108 @@ class WaveguidePort(Port):
+            
+            # Check wether this is manual weighting or string function
+            if E_is_str or H_is_str:
+                if not H_is_str:
```
Thats not a save check? What if H_is_str==True but E_is_str is not?

thliebig commented on python/openEMS/ports.py:
```diff
@@ -1,4 +1,4 @@
-# -*- coding: utf-8 -*-
+	# -*- coding: utf-8 -*-
```
???

thliebig reviewed (changes requested):
In general it would also be nice to eventually add an Matlab/Octave interface and example too? But maybe once this is done and merged...

Does this work for cylindrical coordinates? But as not many (if any) use this is CS, it is  not a deal-breaker...

gadiLhv commented:
Hey @thliebig 

I will review them this week. Seems straightforward enough.

Regarding the Octave/Matlab interface, I will put that on the task list. If you insist, I'll do it now, but I really want to organize my Repo a bit. This on-job git training was very deductive, but I really want to keep it cleaner from here on forward. I'm more fluent in m-scripting than I am in Matlab, so should be simpler.

I need to think about how to do it in cylindrical CS, as I never really used it. I need to figure out the deal with the mesh there. Also, the interpolation schemes I wrote for the mode parser hardly fit that. Add it to the pile, I guess :laughing:.
I Will be happy to sharpen my wits on that. Once the infrastructure is laid down, should be easier.

Hardly a finalized version, but some of my initial mode solver codes are here:
https://github.com/gadiLhv/FEMopenSourceAdventures
The end code will either be in DolfinX or just use a new library called Emerge.

gadiLhv committed:
Fixes per @thliebig requests
Added another example for CPW, propagating in a different direction

gadiLhv commented on FDTD/extensions/operator_ext_excitation.cpp:
```diff
@@ -155,17 +175,28 @@ bool Operator_Ext_Excitation::BuildExtension()
 						if (elec==NULL)
 							continue;
-						if (!elec->GetEnabled())
+						if (!elec->GetEnabled()) // @suppress("Method cannot be resolved")
```
Have no idea how this comment was added. Wish I could program in normal hours.

gadiLhv commented:
Added another example, for CPW. I couldn't verify it against 3rd party software, regretfully, yet.
It's weird that adding PEC blocks makes the convergence worse. I wonder how all the other software do it.

gadiLhv commented:
Here, a nice video

https://github.com/user-attachments/assets/3e679150-8711-49ee-b259-393dc6741799

gadiLhv commented:
Hey @biergaizi 
This is paraview. Simply sliced it down the middle, and added an additional slice at y=0.001

gadiLhv committed:
Fixed mode files in CPW example
Minor changes to test script

gadiLhv commented:
I changed the new test script a bit + made more appropriate mode example files.
Although there is almost no resonance here, I still see very high reflection in the exciting port. I verified that it isn't a bug in how the mode are accepted. I think it's an issue of fringing fields from the waveguide edges, but I have no way to prove this.

<img width="882" height="435" alt="image" src="https://github.com/user-attachments/assets/487362c1-c656-4299-b73c-3943cc7dc33c" />

gadiLhv commented:
Hey @thliebig 

Any other changes necessary for thie PR?

thliebig commented on python/openEMS/ports.py:
```diff
@@ -347,54 +347,123 @@ class WaveguidePort(Port):
+        
+            
+
+            # For the mode file to be used correctly, the direction of 
+            # propagation has to be explicitly set.
+            if not use_function_expr:
```
I think we need to find out?

gadiLhv replied:
If I can remove one redundency, why not. I'd rather deal with this once the CSXCAD issues are resolved, anyway. I'd be happy to test it (crossing fingers that I can). I know for a fact that setting this vector is necessary for everything to work. 
What I don't know, is if the `if` condition is necessary...

biergaizi commented:
My "Pull Request Ganging" proposal has just been implemented. Please try a dummy useless force push to force the CI/CD pipeline to rerun:

    git commit --amend  # without changing anything, just save
    git push --force

After a force push, your openEMS Pull Request will be tested together with your latest CSXCAD Pull Request commits (because both have the same `WaveguideModes` branch name).

thliebig commented:
I can request a re-run manually

gadiLhv commented:
I couldn't figure out what went wrong with the MSVC test. Any hints, @biergaizi & @thliebig?

Nice to see some green lights on this, for a change

biergaizi commented:
> I couldn't figure out what went wrong with the MSVC test. 

Undefined reference due to unexported library functions. On Windows, all symbols are private by default. To make a symbol available to other applications in a DLL, you must mark the class or function as `CSXCAD_EXPORT` (in openEMS, it's `OPENEMS_EXPORT`), which is an alias of `__declspec(dllexport)`.

For example, the class `CSPrimBox` has the signature:

    class CSXCAD_EXPORT CSPrimBox : public CSPrimitives
    {
    // ...
    }

But your `class ModeFileParser` does not export this symbol, so it's unavailable as a public function for openEMS (public in the operating system's `.so`/`.dll` sense, not the C++ sense). This is why you when you try to use `ModeFileParser` from openEMS you get:

```
lld-link: error: undefined symbol: public: void __cdecl ModeFileParser::clearData(void)
>>> referenced by CMakeFiles\openEMS.dir\Common\processmodematch.cpp.obj:(public: virtual void __cdecl ProcessModeMatch::InitProcess(void))

lld-link: error: undefined symbol: public: bool __cdecl ModeFileParser::parseFile(void)
>>> referenced by CMakeFiles\openEMS.dir\Common\processmodematch.cpp.obj:(public: virtual void __cdecl ProcessModeMatch::InitProcess(void))

lld-link: error: undefined symbol: public: void __cdecl ModeFileParser::linInterp2(double, double, double &, double &)
>>> referenced by CMakeFiles\openEMS.dir\Common\processmodematch.cpp.obj:(public: virtual void __cdecl ProcessModeMatch::InitProcess(void))
```

---

But we should be careful here. Everything in `ModeFileParser` will permanently become CSXCAD's public API, and in theory anyone who uses CSXCAD will be able to link against it and use it (even on Unix, because they're public by default). Are you sure that you want everything in `ModeFileParser` to permanently become the public API of CSXCAD?

The class name is not following CSXCAD naming convention of `CSClassName`, and the member function coding style is in `Class.camelCase`, while all CSXCAD's public API uses `Class.CapsForEveryWord`. If this becomes the public API, it will be confusing to downstream users, so I suspect `ModeFileParser` is meant to be an internal function?

I think we have several choices:

1. Full public API integration: Redesign `ModeFileParser` to be a true CSXCAD API, so it's consistent with all other CSXCAD functions.
2. Only expose bare minimum as the public API: Wrap the public part of `ModeFileParser` as `CSModeFileParser`, with just enough functions for the public (including openEMS) to use the mode-matching feature, hide all implementation details.
3. Classify it as an internal API, without future public integration plan, by design: Export `ModeFileParser`, go with the code as-is, but mark `ModeFileParser` as "private to openEMS, do not use" in the documentation to make sure no actual users will call it, perhaps by creating a special namespace, with the future plan to make it remain private in the foreseeable future. 
4. Classify it as an internal API, but open to future public API integration plan: Export  `ModeFileParser`, go with the code as-is, but mark `ModeFileParser` as "experimental, do not use" in the documentation to make sure no actual users will call it, with the future plan to make it a stable part of the public API.

gadiLhv commented:
#1 and #2 are a fundamental change. Both for this and the equivalent PR.
Let me know what to do, @thliebig.

gadiLhv committed:
Changed ModeFileParser to CSModeFileParser

gadiLhv commented:
The fixes here were **very** minor. Applied, as well.

gadiLhv committed:
Initial Commit

Still need to finish the CSXCAD matlab function and port the test script
to matlab.

gadiLhv committed:
Test script not working yet

1. Added the test script
2. Attribute setting not working yet

gadiLhv committed:
Run working. S-Parameter normalization issue

1. Full run successful
2. Minor fixes to script
3. Added propDir property for mode file name

gadiLhv committed:
Cleaned up test script

gadiLhv commented:
Hello again, @thliebig 

A Matlab/Octave script was added to `matlab/examples/transmission_lines`

The m-script functionality is a bit more limited, as it originally supported only TE modes. I matched the same capabilities. 

I translated only one script of the three, but it should be the same for the rest.

tmpusr123 commented:
@thliebig anything more needed here?

thliebig commented:
@gadiLhv I hope not, but I want to do a (hopefully) final code review and testing. But I need to find the time and motivation to do it.

thliebig commented on matlab/examples/transmission_lines/Coax_W_Waveguide_Ports.m:
```diff
@@ -0,0 +1,165 @@
+coax_wire_D = 0.5;
+coax_L = 25
+
+Mode_File_Path = '../../../python/Tests';
```
That is not a good idea. When openEMS gets installed the matlab examples are copied to a "share" folder and this relative path will not work anymore. I think we can just copy these to this example folder instead? I think it is not too bad to have this files redundant for two different purposes?

gadiLhv replied:
Done

thliebig commented on matlab/examples/transmission_lines/Coax_W_Waveguide_Ports.m:
```diff
@@ -0,0 +1,165 @@
+warning off;
+
+addpath('~/opt/openEMS/share/CSXCAD/matlab');
+addpath('~/opt/openEMS/share/openEMS/matlab');
```
These lines should not be here I think. The user should set this in the ".otaverc" or "startup.m" or similar

gadiLhv replied:
Done

thliebig commented on matlab/examples/transmission_lines/Coax_W_Waveguide_Ports.m:
```diff
@@ -0,0 +1,165 @@
+mesh.x = [mesh.x linspace(-(coax_D*0.5 + coax_shield_thick),-coax_D*0.5,4)];
+mesh.x = [mesh.x linspace(-coax_wire_D*0.5,coax_wire_D*0.5,6)];
+mesh.x = [mesh.x linspace(coax_D*0.5,(coax_D*0.5 + coax_shield_thick),4)];
+mesh.x = [mesh.x linspace(coax_wire_D*0.5,coax_D*0.5,7)];
```
I think a homogeneous mesh works best here as this is round?

gadiLhv replied:
Sure. I did this originally to debug some issues I had with fringing fields.. Right now it hardly matters, as long as there is at least one full mesh cell everywhere in the shield.

thliebig commented on matlab/examples/transmission_lines/Coax_W_Waveguide_Ports.m:
```diff
@@ -0,0 +1,165 @@
+Zw = 238.26517157;
+% Line impedance
+Zl = 52.43928218;
+
```
where do these come from? I think this can be calculated from the radii in a closed formula?

gadiLhv replied:
From the mode solver. Not usable for the public, yet.
If you want this from the formula, fine. But can't be done in any case.

gadiLhv replied:
Done

thliebig commented on python/openEMS/ports.py:
```diff
@@ -347,54 +347,123 @@ class WaveguidePort(Port):
+        use_function_expr = None
+        if (self.E_func is not None) and (self.H_func is not None):
+        	use_function_expr = True
+        else:
+        	use_function_expr = False
+        
+        if use_function_expr is None:
+        	raise Exception("Cannot decide if function expression is used or mode file")
```
Does not seem fixed in my latest pull?

thliebig commented on python/openEMS/ports.py:
```diff
@@ -347,54 +347,123 @@ class WaveguidePort(Port):
+                
+                _,Eext = os.path.splitext(self.E_file)
+                _,Hext = os.path.splitext(self.H_file)
+                if not ((Eext == '.csv') and (Hext == '.csv')):
```
also not changed?

thliebig commented:
Post python-Test script do not work anymore due to multiple issues... Please fix them. Thanks

gadiLhv commented:
> Post python-Test script do not work anymore due to multiple issues... Please fix them. Thanks

Can you tell exactly what didn't work? I ran the CSXCAD unit tests, and for me, the update_openEMS.sh script worked.
```
gadi@gadi-virtualbox:~/Repositories/openEMS-Project$ ./update_openEMS.sh ~/opt/openEMS --python --python-venv-mode=disable
enabling Python Extension build
setting install path to: /home/gadi/opt/openEMS
logging build output to: /home/gadi/Repositories/openEMS-Project/build_20260221_223019.log
configuring /home/gadi/Repositories/openEMS-Project via CMake ... please wait
compiling /home/gadi/Repositories/openEMS-Project ... please wait
Building python modules ... please wait
build successful, cleaning up tmp dir ...
 -------- 
openEMS and all modules have been updated successfully...

% add the required paths to Octave/Matlab:
addpath('/home/gadi/opt/openEMS/share/openEMS/matlab')
addpath('/home/gadi/opt/openEMS/share/CSXCAD/matlab')

Have fun using openEMS
```

thliebig commented:
I did not mean the unit tests, I meant the full python demo's. I think they also would have better been added to an examples folder. "python/Tests/Coax_W_WG_Ports.py" and CWP_* is what I meant.

gadiLhv committed:
Code review changes

1. Uniform mesh inside coax
2. Mode files in example path. No relative path redirection.
3. Analytical mode parameter solutions

gadiLhv committed:
Removed the addpath statements

gadiLhv commented:
> I did not mean the unit tests, I meant the full python demo's. 

What should I run to recreate the issues?

> I think they also would have better been added to an examples folder. "python/Tests/Coax_W_WG_Ports.py" and CWP_* is what I meant.

This I understood even less. It's in this folder right now. What would you rather?

thliebig commented:
> What should I run to recreate the issues?

Just run and fix them both? I think I only tested one and there were undefined variables and not up to date function calls.

gadiLhv commented:
Sure, will take care of it tomorrow.

gadiLhv committed:
Minor fixes to test scripts

1. Increase resolution in CPW script
2. Solved variable problems in Coax script

gadiLhv commented:
Solved the issue in the Coax example and improved the performance of the CPW. Also added a disclaimer with things to fiddle with.

gadiLhv commented:
Gute Jahrestag and Happy anniversary to this PR

thliebig commented:
Before I start yet another review on this. Can you confirm that all examples, tests etc. work as they should?

gadiLhv commented:
Ran the Coax and CPW script.
I don't *like* the CPW results, but that's to be resolved with better absorption, in the future. 

Unit tests:

```
gadi@gadi-virtualbox:~/Repositories/openEMS-Project/CSXCAD/python/tests$ python -m unittest test_CSProperties.py 
.............
----------------------------------------------------------------------
Ran 13 tests in 0.121s

OK
gadi@gadi-virtualbox:~/Repositories/openEMS-Project/CSXCAD/python/tests$ python -m unittest test_CSPrimitives.py 
...............
----------------------------------------------------------------------
Ran 15 tests in 0.014s

OK
gadi@gadi-virtualbox:~/Repositories/openEMS-Project/CSXCAD/python/tests$ python -m unittest test_CSXCAD.py 
..
----------------------------------------------------------------------
Ran 2 tests in 0.002s

OK
```

thliebig commented:
A modified version has been pushed, the transmission line examples I left out for now as the waveguide probes are very confusing for a user as that does not at all give anything close to what a VNA would measure. 
For now mode profile probes should only be used for waveguides until we find a better way to adept to transmission lines..
